### DISCUSSION TITLE
SPAM: A little research

### DISCUSSION THREAD

**Discussion by shleeable** [mastodon/mastodon#18970] [General]
# SPAM: A little research

## Problem statement
Spammer X can join Instance Y. Spam SEO trash, Get suspended from instance Y, then join instance Z because there is no shared knowledge of this threat actor in the AP ecosystem.

A lot of spam is done by humans... but if you limit the spam in by 20% with a Captcha, you can catch the next 20% with existing spam block lists (like stopforumspam)... and the next 20% with a federated suspend relay. Don't think it's an impossible problem to block MOST of the spam.

## What can we do about it?

I believe "we" should aim to add 3 different kinds of anti-spam as a defense in depth method
1) Captcha
2) Blocklist Relay (API Endpoint)
3) Instance level protections

*The real problem is signal to noise, we're the fediverse. You need to build something for the fediverse with fediverse data.*

An existing third party anti-spam ecosystem is built on email, and forum spam... I believe we should build similar SAAS style services dedicated to AP with new fresh data taken from live ActivityPub instances, and use these existing services offering additional metrics.

For starters, free third party APIs stopforumspam.com and BotScout.com should be considered.

### Captcha
A `questyCaptcha` as used wikipedia (unique questions per instance.. "what is 1 +1?") or something locally generated
or 
something closer to a recaptcha-ish service 
* https://friendlycaptcha.com/ looks good... 
* https://mcaptcha.org/ (selfhosted is excellent, so I'd like to investigate this further).

### Blocklist relay
Original quote from Discord `the proof of concept thing I wanted to build was a Spam/suspension "relay" (not the same as current relays but a similar concept).... A bunch of trusted instances can sign up, share threat info and hopefully block spammers together... On signup, my instance has request if the ip and email has been suspended previously.`

This is what I've talked about previously.. make a "antispam relay", instance admins can opt in to join, and when an instance admin suspends a spammer, that spam is forwarded to the relay. When a spammer joins another instance, their username/ip/email is checked against the blocklist of previous spammers.

This is good because additional checks can be added the antispam relay to fight spammers, without having to add these methods to all of the instances using the relay. offering real-time passive protection.

How this would work in practice to limit false positives.
* On signup, if there is no history of spam attached to the email or ip... business as usual 
* if the signup is attached to spam, put it in the manual review queue.

### Instance level protections
MRF or similar to block new users posting links. Obviously this is easy to bypass, but in a defense in depth style, you need to add multiple bottlenecks.
https://git.pleroma.social/pleroma/pleroma/-/merge_requests/1307/diffs
https://git.pleroma.social/pleroma/pleroma/-/merge_requests/3278
http://www.corpit.ru/mjt/rbldnsd.html

You should block the TOR relays with this list https://check.torproject.org/exit-addresses

* Wikipedia maintain a large list of spam domains and spam email (Let's not copy/paste but we could make our own)
https://github.com/wikimedia/mediawiki-extensions-SpamBlacklist
https://www.mediawiki.org/wiki/Extension:SpamBlacklist
Data: https://meta.wikimedia.org/wiki/Spam_blacklist 
Data: https://en.wikipedia.org/w/index.php?title=MediaWiki:Spam-blacklist

---

More Examples: Wikipedia uses a series of plugins to limit spam
https://www.mediawiki.org/wiki/Manual:Combating_spam
https://www.mediawiki.org/wiki/Spam_Filter

* Blocking open proxies / TOR from anonymous editing
* Use of Captcha for signups/edits (Wikipedia recommend the use of Question Captchas)
https://www.mediawiki.org/wiki/Extension:ConfirmEdit
https://www.mediawiki.org/wiki/Extension:QuestyCaptcha

---

In summary, The cat and mouse game will never stop. that's why you kind of need do multiple things. AP is magic, and we can utilize the thousands of instance admins to work together and built protection against spam. 

---

Related reading
Email spam makes heavy use of DNSBLs (DNS-based blacklists) https://www.spamhaus.org/sbl/
https://en.wikipedia.org/wiki/Comparison_of_DNS_blacklists

There are common services which could be used or emulated
https://akismet.com/ (Wordpress spam)
https://www.spamcop.net/ (email)
https://www.oopspam.com/ (paid) (edited)

**Comment by shleeable** [comment]
I think the question/reply is an easy win, and it's opt-in. I could think of 30 unique questions that won't be too hard for international people for my instance.

I mean, the question/reply method has that flaw that people would just make a list of instance questions with the answers.. but that means they need to do a lot of manual work

Freeform text with multiple answers is best

but if spammers are going to put in the effort to make a list of questions and answers... something like this would be fine

![image](https://user-images.githubusercontent.com/17537000/184540746-02543ec2-b70e-4e66-b982-0713f3e3b17e.png)

As long as I can create multiple questions/answer combos for my instance

but also.. "what's the title of this film? with a free text"
that's a nice compromise to me?
"dead poets" "dead poets society" "dead poet's society" accepted
The visual question causes problems with accessibility.
Nothing is perfect. so you need multiple solutions.

**Reply by gilgongo** [reply]
Hi - I would **definitely** support the implementation of a blocklist relay endpoint and various instance level protections. But I'm afraid captcha needs to die in a fire. Unless it's only shown to a very, very selective subset of users, annoying the vast majority of good actors for the sake of a tiny number of bad ones is not a good idea at all.

**Comment by shleeable** [comment]
notes: 

```
I think all of this is about adding friction and frustration.
if they have to spend more and more time to create accounts.
they'll get more annoyed once we block them.

hat's why I think the second anti-spam relay idea makes sense to me. I figure this isn't a hard problem if done well..

I'd imagine at the start, the block relay would need to approve instances when they want to join - and if one of the members starts being a bad actor and sending fake suspends, kick them and remove all of their data.

As the block relay grows, the instances on the relay should be trusted to a certain extent... like I'd trust suspends from all of the instances in my region for example. That's 5 of us out of the gate.

Following the AP way, people should be able to join multiple relays to get more coverage.

in the same way I've blocked spinster.. I wouldn't join a blockrelay with that instance on it

and finally, I figure when somebody signs up to my instance, and the block relay has alerted the email is bad... that signup should just be moved to the manual approve queue (not blocked) due to false positives. 
and has for PII.. hash the email, hash the username.. ip is ip... you're only storing PII from bad actors. which are likely not using real emails, usernames etc.

I think spa mentioned dropping tor exit nodes limited their spam exposure.. so having signups from tor ips going to the manual approval because the block relay is maintaining a list of tor exit nodes for me.. sounds pretty cool?
```

**Comment by shleeable** [comment]
This is just an example
but most spammers are also spamming in multiple places. I had a spammer on my mastodon and I found them on stopforumspam.com. same email, similar username
![image](https://user-images.githubusercontent.com/17537000/186057168-d4de97be-d83c-4104-abd9-7b9136b87583.png)

![image](https://user-images.githubusercontent.com/17537000/184568885-6a532d41-4a06-45ff-8974-5839a44ed5cb.png)


Anti-spam for PHPBB or wikipedia or even email has existed for ages.. the threat modals are all well known.
Masto already has tools to block IPs and emails. I just want to share the threat inteligence between administrators.

**Comment by shleeable** [comment]
https://developers.google.com/safe-browsing/
https://urlhaus.abuse.ch/browse/
https://docs.ipdata.co/docs/blocklists
https://honeydb.io/
crowdsec

**Comment by stopforumspam** [comment]
I'm more than willing to work with anyone on this, just let me know.

if you want to bake in stopforumspam support but need something from me then I'll make that happen
